src/NLog/Targets/Target.cs

```diff
@@ -722,6 +724,39 @@ protected string RenderLogEvent(Layout layout, LogEventInfo logEvent)
             }
         }
 
+        /// <summary>
+        /// Renders the event info in layout.
+        /// </summary>
+        /// <param name="layout">The layout.</param>
+        /// <param name="logEvent">The event info.</param>
+        /// <param name="defaultValue">The default value when <paramref name="layout"/> and/or <paramref name="logEvent"/> is <c>null</c></param>
+        /// <returns>Value representing log event.</returns>
+        protected T RenderLogEvent<T>([CanBeNull] Layout<T> layout, [CanBeNull] LogEventInfo logEvent, T defaultValue = default(T))
+        {
+            if (layout == null || logEvent == null)
+                return defaultValue;
+
+            if (OptimizeBufferReuse)
+            {
+                if (layout.IsFixed)
+                {
+                    return layout.RenderToValueOrDefault(logEvent, defaultValue);
+                }
+
+                if (TryGetCachedValue(layout, logEvent, out var value))
+                {
+                    return value;
+                }
+
+                using (var localTarget = ReusableLayoutBuilder.Allocate())
+                {
+                    return layout.RenderToValueInternal(logEvent, localTarget.Result, default(T));
```

Comment:
When finally ending up at using `ReusableLayoutBuilder` to render into string-output, and then convert into the typed-value, then it would be nice if the caching logic could avoid performing StringBuilder.ToString() but instead check if the _cachedKey (as string) matched the contents of StringBuilder using the EqualTo-method (So returned the cached converted value without first allocating a temporary string-object)

Reply:
I prefer to do nice to haves after merge. This PR is too long active

Reply:
Feel free to send a PR for this after this one get merged